Add comprehensive CRD versioning flow to documentation (#16026)
kubernetes/website: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning.md

{{< feature-state state="beta" for_kubernetes_version="1.15" >}}

The CustomResourceDefinition API provides a workflow for introducing and upgrading
to new versions of a CustomResourceDefinition.

When a CustomResourceDefinition is first created, the first version is set at
an appropriate stability level and an version number. For example `v1beta1`
would indicate that the first version is not yet stable. All custom resource objects
will initially be stored at this version.

Once the CustomResourceDefinition is created, clients may begin using the
`v1beta1` API.

Later it might be necessary to add new version such as `v1`.

When adding a version to the CustomResourceDefiition, custom resource objects
will need to be served at both versions. This means that they will sometimes be
served at a different version they are stored at. In order for this to be
possible, the custom resource objects must sometimes be converted between the

> version they are stored at and the version they are served at. How custom

Reviewer: (applies in many places) `storage version` (explained once at first use) might read more easily than `version they are stored at`. similarly, `served version` might read more easily than `version they are served at`

resource objects are converted is determined CustomResourceDefinition's

> conversion strategy. If the conversion involves schema changes and requires

Reviewer: ```suggestion
conversion strategy. If the conversion involves schema changes or requires
```

custom logic, a conversion webhook should be used. Once a conversion strategy

> decided on, and all conversion webhooks are deployed, the

Reviewer: ```suggestion
decided on, and all conversion webhooks are deployed and available, the
```

CustomResourceDefinition can be updated to serve both the old and new versions.

> Once there are two versions, the storage version may be change to the new

Reviewer: ```suggestion
Once there are two versions, the storage version may be changed to the new
```

version at any time to begin the migration of stored custom resource objects to

> the new version. Changes to the storage version are "lazy" and Custom resources

Reviewer: rephrase, maybe to `Changes to the storage version take effect for each custom resource instance the next time that resource is written.`

will continue be stored at the old version until they are next written to.

> Next, clients may incrementally migrated to the `v1` new version. It is perfectly

Reviewer: make it clear clients can start using the new version before/after/during the storage migration

Author: I've expanded on the storage migration and also incorporated this suggestion

safe for some clients to use `v1beta` while others use `v1`.

Eventually the old version may need to be removed. Before removing a version,
first ensure all clients are fully migrated to the new version and that the
CustomResourceDefinition is configured to store custom resource objects at the
new version. Next, make sure all stored custom resource objects have be migrated
to a new version, either by using the [Storage Version
Migrator](https://github.com/kubernetes-sigs/kube-storage-version-migrator) or

> by manually writing to each custom resource object at least once.

Reviewer: > by manually writing to each custom resource object at least once.

make it clear a no-op write `get -> put identical content` is sufficient... no modifications are required

Author: Thanks, I've incorporated this

> Finally, old version can then be removed and converter support for the old

Reviewer: ```suggestion
Finally, the old version can be removed from the CustomResourceDefinition, and converter support for the old
```

{{< feature-state state="beta" for_kubernetes_version="1.15" >}}

The CustomResourceDefinition API provides a workflow for introducing and upgrading
to new versions of a CustomResourceDefinition.

When a CustomResourceDefinition is first created, the first version is set at
an appropriate stability level and a version number. For example `v1beta1`
would indicate that the first version is not yet stable. All custom resource objects
will initially be stored at this version.

Once the CustomResourceDefinition is created, clients may begin using the
`v1beta1` API.

Later it might be necessary to add new version such as `v1`.

Adding a new version:
1. Select a conversion strategy. Since custom resource objects need to be able

>    to be served at both versions, that they will sometimes be served at a

Reviewer: ```suggestion
   to be served at both versions, that means they will sometimes be served at a
```

>    different version they are stored at. In order for this to be possible, the

Reviewer: ```suggestion
   different version than their storage version. In order for this to be possible, the
```

   custom resource objects must sometimes be converted between the version they
   are stored at and the version they are served at. If the conversion involves
   schema changes and requires custom logic, a conversion webhook should be
   used. If there are no schema changes, the default `None` conversion strategy

>    may be used.

Reviewer: ```suggestion
   may be used, and only the `apiVersion` field will be modified when serving different versions.
```

2. If using conversion webhooks, create and deploy the conversion webhook. See
   the [Webhook conversion](#webhook-conversion) for more details.

> 3. Update the CustomResourceDefinition to serve both the old and new versions

Reviewer: tell how to do this specifically (e.g. "... by adding the new version to the `spec.versions` list with `served:true`")

Author: Thanks. I've done a pass over the doc changes to spell out what exact fields to change for each step that requires a field change.

   and to use the selected conversion strategy. If using a conversion webhook,
   also set the webhook configuration.

> Once the new version is added, clients may incrementally migrated to the `v1`

Reviewer: ```suggestion
Once the new version is added, clients may incrementally migrate to the
```

new version. It is perfectly safe for some clients to use `v1beta1` while others
use `v1`. It is safe for clients to use both versions before, during and after
the storage migration.

Storage Migration:
1. Update the version the custom resources objects are stored at in the

>    CustomResourceDefinition to the new version. Changes to the storage version

Reviewer: tell how to do this specifically (e.g. "... by changing the `storage` attribute of the old version to `false` and the `storage` attribute of the new version to `true`)

   take effect for each custom resource instance the next time that resource is
   written.

> 1. Complete the storage migration

Reviewer: ```suggestion
1. Migrate the existing instances
```

  - Option 1: Use [Storage Version Migrator](https://github.com/kubernetes-sigs/kube-storage-version-migrator)
  - Option 2: Write to each custom resource object at least once. A get followed by a put of the identical

>     content is sufficient. No modification of the data is required.

Reviewer: step 3 of the migration (after ensuring all existing instances were persisted in the new version) is removing all other versions from the CustomResourceDefinition `status.storedVersions` field (patching the custom resource definition `status` subresource with a merge patch of `{"status":{"storedVersions":["v1"]}}` would be one way to do this

Reviewer: > removing all other versions from the CustomResourceDefinition `status.storedVersions` field

@caesarxuchao does the storage migrator intend to handle doing this?

Author: I've referenced the `#upgrade-existing-objects-to-a-new-stored-version` section here instead of listing out the steps in detail. I've also updated the `#upgrade-existing-objects-to-a-new-stored-version` to include the storage migrator as an option.

Removing an old version:
1. Ensure all clients are fully migrated to the new version. The kube-apiserver
   logs can reviewed to help identify any clients that are still accessing via the
   old version.
1. Ensure CustomResourceDefinition is configured to store custom resource objects at the new version
1. Ensure the storage migration, described above, has been completed. If using a

>    conversion webhook, check if it is recieving any conversions requests when

Reviewer: ```suggestion
   conversion webhook, check if it is receiving any conversions requests when
```

   making list requests for custom resources at the old version. If not, all

>    custom resource objects have been converted.

Reviewer: ensure the old version is no longer listed in the CustomResourceDefinition `status.storedVersions`

> 1. Remove the old version from the CustomResourceDefinition

Reviewer: be specific here (e.g. "... from the CustomResourceDefinition `spec.versions` list")
